### PR TITLE
ICU-13417 Add the Locale::(get|set)(Unicode)?KeywordValue() functions.

### PR DESCRIPTION
**Pull request by roubert**
Additional functions for setting and getting both legacy and Unicode
keywords and values.

### CODE REVIEW

File: icu4c/source/common/unicode/locid.h [hunk 1]
```diff
@@ -523,6 +577,54 @@ class U_COMMON_API Locale : public UObject {
      */
     void setKeywordValue(const char* keywordName, const char* keywordValue, UErrorCode &status);
 
+#ifndef U_HIDE_DRAFT_API
+    /**
+     * Sets or removes the value for a keyword.
```
Comment (markusicu) [inline icu4c/source/common/unicode/locid.h]: On both the existing and new keyword value setters, please add documentation as discussed in the API discussion.
https://docs.google.com/document/d/15r3s4ZUFw1crAK44VHi6UPbvrXQca9b-nFRCyFKKzHc/edit#heading=h.dwuzemgnou2m

"set[Unicode]KeywordValue() document no mapping vs. accept Unicode names/values"

File: icu4c/source/common/unicode/locid.h [hunk 2]
```diff
@@ -523,6 +577,54 @@ class U_COMMON_API Locale : public UObject {
...
+     *
+     * @param keywordName name of the keyword to be set.
+     * @param keywordValue value of the keyword to be set. If 0-length or
+     *  NULL, will result in the keyword being removed. No error is given if
+     *  that keyword does not exist.
+     * @param status Returns any error information while performing this operation.
+     *
+     * @draft ICU 63
+     */
+    void setUnicodeKeywordValue(const char* keywordName, const char* keywordValue, UErrorCode& status);
```
Comment (markusicu) [inline icu4c/source/common/unicode/locid.h]: You did not propose this version. We should not add what's not proposed. Do we need it?

File: icu4c/source/common/unicode/locid.h [hunk 3]
```diff
@@ -507,6 +507,60 @@ class U_COMMON_API Locale : public UObject {
...
+
+    /**
+     * Gets the Unicode value for a Unicode keyword.
+     *
+     * @param keywordName  name of the keyword for which we want the value.
+     * @param status       error information if getting the value failed.
+     * @return             the keyword value.
+     * @draft ICU 63
+     */
+    const char* getUnicodeKeywordValue(const char* keywordName, UErrorCode& status) const;
```
Comment (markusicu) [inline icu4c/source/common/unicode/locid.h]: You did not propose this version. We should not add what's not proposed. Do we need it?

Also, it seems like this version may have issues with returning temporary memory.

File: icu4c/source/common/locid.cpp [hunk 4]
```diff
@@ -1202,6 +1203,130 @@ Locale::getKeywordValue(const char* keywordName, char *buffer, int32_t bufLen, U
...
+    }
+
+    // TODO: Remove the need for a const char* to a NUL terminated buffer.
+    const CharString keywordName_nul(keywordName, status);
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    LocalMemory<char> scratch;
+    int32_t scratch_capacity = 8;  // Arbitrarily chosen default size.
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp]: I would start bigger, for example with 12 or 16.
In BCP 47, subtags can be up to 8 long, and some legacy values are longer (e.g., "traditional").

Comment (markusicu) [inline icu4c/source/common/locid.cpp]: Please use camelCase variable names like in Java.

File: icu4c/source/common/locid.cpp [hunk 5]
```diff
@@ -1202,6 +1203,130 @@ Locale::getKeywordValue(const char* keywordName, char *buffer, int32_t bufLen, U
...
+    {
+        CharStringByteSink sink(&legacy_value);
+        getKeywordValue(legacy_key, sink, status);
+    }
+
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+
+    const char* unicode_value = uloc_toUnicodeLocaleType(
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp]: Note that unicode_value can be the same as legacy_value.data() which will become invalid. http://icu-project.org/apiref/icu4c/uloc_8h.html#a0261e5fbb0ea96e16d111b681426f2a4

Luckily you didn't propose this particular function so you need not/should not implement it...

File: icu4c/source/common/unicode/locid.h (hunk 2 above)
Comment (markusicu) [reply icu4c/source/common/unicode/locid.h]: We didn't record anything about this function. Let's drop it.

File: icu4c/source/common/unicode/locid.h (hunk 3 above)
Comment (markusicu) [reply icu4c/source/common/unicode/locid.h]: ditto

File: icu4c/source/common/unicode/locid.h (hunk 2 above)
Comment (roubert) [reply icu4c/source/common/unicode/locid.h]: Done.

File: icu4c/source/common/unicode/locid.h (hunk 3 above)
Comment (roubert) [reply icu4c/source/common/unicode/locid.h]: Done.

**Pull request merged**